### PR TITLE
Pass through arbitrary attributes to QuickGrid

### LINKED ISSUE #45327: Extend Quickgrid with Id and Style Properties
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

yes, id property is important to set for automated-testing like bunit test or cypress/playwright/seleniu, etc.

### Describe the solution you'd like

it is not possible to set id and style for quickgrid table from outside. I would like to see an **id** for quickgrid that can be used for automated-testing
There is **Class** property however **Style** would be also useful in order to use it in-style css directly without using a css-class

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ElderJames**
# Pass through arbitrary attributes to QuickGrid

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

Add the `AdditionalAttributes` parameter with `CaptureUnmatchedValues` flag and pass through to QuickGrid table element.

## Description

Follow up https://github.com/aspnet/AspLabs/pull/559#issuecomment-1414094517

Fixes #45327

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
Thanks @ElderJames! This is looking really good.

One small issue will arise with this implementation, though. The grid already generates a `class` attribute automatically adding both `quickgrid` and a `loading` class that toggles based on any associated `ItemsProvider`. The implementation here will overwrite that if the developer provides any `class` attribute, thereby breaking the "loading" indicator and other styles.

To solve this, I would suggest:

 * When the grid is rendered at https://github.com/dotnet/aspnetcore/pull/48268/files#diff-6d509929f1d48812b35dbac8a474371c00e27c0932219e323dc11cfff82ea460R13, move the `class="@GridClass()"` to be after the `@attributes`. Then `GridClass` will always take precedence.
 * Change the implementation of `GridClass` so that, if the developer has supplied `AdditionalAttributes`, it looks for any `class` entry in that dictionary and appends it to the `GridClass` result value.

This would preserve existing behaviors while also letting the developer add extra classes using a normal `class` attribute.

**Comment by ElderJames** [comment]
Thanks for your suggestions @SteveSandersonMS , I have fixed this. But I'm a little confused about the test and why the `style` won't be passed to the table.

**Comment by ElderJames** [comment]
Hello @SteveSandersonMS , could you please teach me how to solve the style attribute issue?

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs [hunk 1]
```diff
@@ -387,7 +392,15 @@ private string AriaSortValue(ColumnBase<TGridItem> column)
         : ColumnClass(column);
 
     private string GridClass()
-        => $"quickgrid {Class} {(_pendingDataLoadCancellationTokenSource is null ? null : "loading")}";
+    {
+        var gridClass = $"quickgrid {Class} {(_pendingDataLoadCancellationTokenSource is null ? null : "loading")}";
+        if (AdditionalAttributes?.TryGetValue("class", out var originalClass) == true)
+        {
+            gridClass += $" {originalClass}";
+        }
```
Comment (mkArtakMSFT) [inline src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs:400]: You can utilize the AttributeUtilities.CombineClassNames method instead: https://github.com/dotnet/aspnetcore/blob/33243ad0a057a1d99fd2b7fe0f9765d4b28d641d/src/Components/Web/src/Forms/AttributeUtilities.cs#L10

Here is a usage example in the `CssClass` property of the `InputBase` type; https://github.com/dotnet/aspnetcore/blob/33243ad0a057a1d99fd2b7fe0f9765d4b28d641d/src/Components/Web/src/Forms/InputBase.cs#L187

**Review by mkArtakMSFT — CHANGES_REQUESTED** [review]
Thanks for your PR.
Glad to see updated tests too. There is just one minor optimization I've left a comment about and then I think this will be in good shape to take.

File: src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs (hunk 1 above)
Comment (ElderJames) [reply src/Components/QuickGrid/Microsoft.AspNetCore.Components.QuickGrid/src/QuickGrid.razor.cs]: Thank you @mkArtakMSFT , I have done the changes, but AttributeUtilities.CombineClassNames is internal that can't be accessed in QuickGrid library. How to solve that?

**Comment by SteveSandersonMS** [comment]
Thanks for the update. This is now ready to merge, except for fixing the final compiler issue. We need it to go in today, so I've made the necessary change and am continuing the PR in https://github.com/dotnet/aspnetcore/pull/50051, so will close this PR in favour of that. The new PR still uses the original commits from here so hopefully @ElderJames will still be credited as the author.

**Pull request closed without merging**